unit/src/InputParametersTest.C

```diff
@@ -259,3 +259,12 @@ TEST(InputParameters, makeParamRequired)
         << "failed with unexpected error: " << msg;
   }
 }
+
+TEST(InputParameters, defaultPPValues)
+{
+  // programmatically set default value of PPName parameter
+  InputParameters p1 = emptyInputParameters();
+  p1.addParam<PostprocessorName>("pp_name", "testing");
+  p1.set<PostprocessorName>("pp_name") = "1";
```

Comment:
I wouldn't expect this approach to work. The default values are supposed to be set at the time the parameter is added. The "set" method is used for both input file setting or programmatic setting of a parameter. I guess I hadn't anticipated the need for a "programmatic setting of the default". We could probably handle this in one of two ways: 1) We "re-add" the parameter again matching the name and type and supplying the default. 2) We add a new method to InputParameters that specifically used for "programmatic defaults". 

So first off, do you need this? Maybe I'm not seeing the use case but it seems that the default should be supplied when the parameter is first declared. If you have a case then we can consider one of the solutions above.

Reply:
I don't need this and I don't see a great use for me in the future. However, casual contributors to MOOSE may assume this to work and be confused if it does not. If you don't see an issue with not supporting programmatic defaults, I won't worry about it.